### PR TITLE
fix: re-write index imports using TypeScript's resolved module path

### LINKED ISSUE #458: Index import rewriting broken
Two scenarios:
package.json has a main field which points to another spot.

`node_modules/foo/package.json`:
```
"main": "dist/index.js",
"typings": "dist/index.d.ts",
```
`user.ts`:
```
import * as f from 'foo';
```

rewrites as 
```
import * as f from 'foo/index';
```
it should instead be 
```
import * as f from 'foo/dist/index';
```

Also, 
```
import {a} from './';
```
with an index.ts in the same directory produces
```
import {a} from './/';
```

/cc @frozenpandaz

### PR DESCRIPTION
**Pull request by alexeagle**
fixes #458

### CODE REVIEW

File: src/tsickle.ts [hunk 1]
```diff
@@ -469,6 +470,8 @@ class ClosureRewriter extends Rewriter {
 type HasTypeParameters =
     ts.InterfaceDeclaration|ts.ClassLikeDeclaration|ts.TypeAliasDeclaration|ts.SignatureDeclaration;
 
+const EXT = /(\.ts|\.d\.ts|\.js|\.jsx|\.tsx)$/;
```
Comment (evmar) [inline src/tsickle.ts]: Doc or give a better variable name?  Not sure what this is for.

File: src/tsickle.ts [hunk 2]
```diff
@@ -853,13 +856,20 @@ class Annotator extends ClosureRewriter {
       }
       const resolved = ts.resolveModuleName(moduleId, this.file.fileName, this.tsOpts, this.host);
       if (resolved && resolved.resolvedModule) {
-        const resolvedModule = resolved.resolvedModule.resolvedFileName.replace(/(\.d)?\.ts$/, '');
-        const requestedModule = moduleId.replace(/\.js$/, '');
+        const requestedModule = moduleId.replace(EXT, '');
+        const resolvedModule = resolved.resolvedModule.resolvedFileName.replace(EXT, '');
         // If the imported module resolves to foo/index, but the specified module was foo, then we
         // append the /index.
```
Comment (evmar) [inline src/tsickle.ts]: I think this comment is obsolete now?

File: src/tsickle.ts [hunk 3]
```diff
@@ -853,13 +856,20 @@ class Annotator extends ClosureRewriter {
...
-          moduleId += '/index';
+        if (requestedModule.substr(requestedModule.lastIndexOf('/')) !==
+            resolvedModule.substr(resolvedModule.lastIndexOf('/'))) {
+          moduleId = path.relative(this.file.fileName, resolvedModule)
+                         // path.relative starts from the parent directory (../self) but ES6
+                         // imports are relative to the current directory (./self) so we strip
+                         // one dot.
+                         .replace(/^\.\./, '.')
+                         // moduleResolution=node support: if importing from under
+                         // node_modules, omit the leading part of the path.
```
Comment (evmar) [inline src/tsickle.ts]: Will this get confused by the real node_modules dir in g3?

File: src/tsickle.ts [hunk 4]
```diff
@@ -853,13 +856,20 @@ class Annotator extends ClosureRewriter {
...
         // append the /index.
-        if (resolvedModule.substr(resolvedModule.length - 6) === '/index' &&
-            requestedModule.substr(requestedModule.length - 6) !== '/index') {
-          moduleId += '/index';
+        if (requestedModule.substr(requestedModule.lastIndexOf('/')) !==
+            resolvedModule.substr(resolvedModule.lastIndexOf('/'))) {
+          moduleId = path.relative(this.file.fileName, resolvedModule)
+                         // path.relative starts from the parent directory (../self) but ES6
+                         // imports are relative to the current directory (./self) so we strip
+                         // one dot.
```
Comment (evmar) [inline src/tsickle.ts]: I don't quite follow this.  Why does it start from the parent directory?

Comment (alexeagle) [inline src/tsickle.ts]:
```
$ node
> path.relative('a/b', 'a/c')
'../c'
```
but we would like to import `./c` from `a/b`
How could I explain it better?

File: src/tsickle.ts (hunk 3 above)
Comment (alexeagle) [inline src/tsickle.ts]: We don't even run this code in google3 - the `convertIndexImportShorthand` option is set to false.

This only matches strings where TypeScript actually resolved a module. If google3 TypeScript were to resolve a file under a node_modules directory, we'd be in trouble already.

File: src/tsickle.ts (hunk 4 above)
Comment (evmar) [inline src/tsickle.ts]: Oh I see, it treats the first argument as if it's a directory.  Maybe it'd be more readable as
```js
path.relative(path.dirname(this.file.fileName), resolveModule)
```
and then drop the regex??

Comment (alexeagle) [inline src/tsickle.ts]: Sure, though I still need to add a `./` on the front to keep it relative
```
> path.relative(path.dirname('a/b'), 'a/c');
'c'
```

**Comment by alexeagle** [comment]
Turns out this was the wrong fix.

Closure recently supports reading the `package.json` file so we don't need to compensate for node_modules which have a `main` field.
At the same time, this does the wrong thing - it uses the layout of files on the type side (where is the .d.ts file) instead of the value side (where is the .js file). These layouts are typically different, eg. the Jscookie example linked above.

The only case where we should do index rewrites is when the import is implicitly from `/index.js`.

**Comment by alexeagle** [comment]
reverting in https://github.com/angular/tsickle/pull/472

**Pull request merged**